### DISCUSSION TITLE
Populating Entity Dictionary (Elide 6)

### DISCUSSION THREAD

**Discussion by Asky-GH** [yahoo/elide#2533] [Q&A] [answered]
Hi!
I'm not 100% sure if it's a bug but currently, the entity dictionary population mechanism works in an unpredictable way.
Suppose I have such entities
```java
@Include(name = "posts")
@Entity
public class Post {

    @Id
    private UUID id;

    @Embedded
    private Audit creation;

    @Embedded
    private Audit modification;
}

@Embeddable
public class Audit {

    @ManyToOne
    private User user;

    private LocalDateTime dateTime;
}

@Include(name = "users")
@Entity
public class User {

    @Id
    private UUID id;
}
```
Since entities are being bound in a random order, if it happens that the **Post** entity is bound ahead of the **User** entity, then the **/users** endpoint won't be exposed.
Сould you please comment on this behavior?

**Comment by aklish** [comment]
That sounds like a bug.  I've never seen that behavior.  What data store are you using?  Are you using the multiplex store?  Can you reproduce this with a small project based on elide-spring-boot-example?

**Reply by Asky-GH** [reply]
Here you are
Since the order in which entities are being bound is important to reproduce such behavior, you can put the breakpoint in the **JpaDataStore**
```java
// Use the entities defined in the entity manager factory.
for (EntityType type : entityManagerSupplier.get().getMetamodel().getEntities()) {
```
and evaluate the expression **entityManagerSupplier.get().getMetamodel().getEntities()**
Restart the application until **Post** will come before **User**
[elide.zip](https://github.com/yahoo/elide/files/8026541/elide.zip)

**Comment by aklish** [comment]
Thanks.  I can reproduce the issue.  There is a bug here for sure.  Will provide an update soon.

**Comment by aklish** [comment]
I did a bit of digging.  The problem is that there is an Elide mode (Post) which has an embedded attribute (Audit)  which contains a relationship to another model (User).  

From a JSON-API perspective - attributes and relationships are two different concepts.  An attribute does not reference/contain a relationship.   There is no way to express this in the API (even though JPA allows this kind of modeling).  

What you can do is to pull properties from the User model into the Audit embedded attribute to get a similar effect.  Let me know if this makes sense.

**Reply by Asky-GH** [reply]
Does it mean that it's expected behavior and it won't be fixed?

**Reply by Asky-GH** [reply]
And if you are just proposing some temporary solution and planning to fix such behavior then it's not suitable for my project. Currently, I ended up adding **Exclude** on embedded objects.

**Comment by aklish** [comment]
Elide treats embedded bindings different from standard model bindings.  An ID field is not required and they are not exposed as API endpoints.   When Elide processes Post first, it discovers the embedded attribute Audit.   It then attempts to bind the hierarchy of classes that make up that attribute.   In the process of doing this, it discovers User and binds it as an embedded model.   Later when the User model is bound by the JPA store, it discovers that it is already bound (this is why it is not exposed).

There are two behaviors we could explore:
1. Throwing an exception if an Embedded model has a relationship or has the Include annotation.   Embedded models cannot have relationships as there is no equivalent in JSON-API to represent them.
2. We could attempt to separate the attribute and model bindings so a regular elide model can also be bound as an attribute model and not conflict.   The concern I have with this approach is cycles and large relationship hierarchies.  If an embedded model has a relationship to another model, and that in turn has relationships to other models, where do we stop the embedding?  There is no way to prune this tree.  This is why I believe the model is the root of the problem.

My suggestion is to create a separate class with the subset of User properties you want to expose - and make that embedded.

**Reply by Asky-GH** [reply]
Isn't it possible to postpone embedded bindings?

**Reply by aklish** [reply]
Another possibility is we could skip (instead of throwing an error or silently failing) any entities that are elide models in embedded attributes.  That would allow the developer to expose some of those attributes of the embedded model as computed attributes.  Thoughts on this approach?

**Reply by Asky-GH** [reply]
Sounds even better.

**Reply by aklish** [reply]
Ok.  Will try to get this in the next release.

**Comment by aklish** [comment, marked as answer]
Here is a patch.  I need to add a test, but you can try it out if you want.

https://github.com/yahoo/elide/pull/2535

**Reply by Asky-GH** [reply]
When the release is scheduled. Don't know if I can figure it out quickly (what kind of test/what are the conventions of writing them).

**Reply by aklish** [reply]
Probably next week (usually Friday).

**Reply by aklish** [reply]
Might get one out tomorrow.

**Reply by aklish** [reply]
Release 6.1.2 has the fix.

**Reply by Asky-GH** [reply]
Thanks! Looks like the issue is solved but it seems like the new version of Elide has a bug related to GraphQL.
Currently, we are not using GraphQL in our project but after the update, looks like such configuration doesn't disable it:
```java
elide:
  json-api:
    path: /
    enabled: true
  graphql:
    enabled: false
```
We've ended up using the following solution but I'm not sure how robust is it and what consequences will it lead to:
```java
    @Bean
    public QueryRunners createQueryRunners() {
        return null;
    }
```

**Reply by aklish** [reply]
In the elide-spring-boot-example project, when I set 'enabled' to false, I get the following error:

```
{
  "timestamp": "2022-02-28T16:39:18.919+00:00",
  "status": 404,
  "error": "Not Found",
  "path": "/graphql/api/v1"
}
```

This is with Elide 6.1.2.  Are you using the same version of Elide throughout your project?

**Reply by Asky-GH** [reply]
Hi @aklish!
Sorry for such a late response.
To reproduce this issue try changing the endpoint of **ArtifactGroup** from _group_ to _artifact-group_.

**Reply by aklish** [reply]
I still can't reproduce.  There must be something environment specific that is causing this.
